src/Shared/PackageManagers/ProjectManagerFactory.cs

```diff
@@ -8,55 +8,153 @@ namespace Microsoft.CST.OpenSource.PackageManagers
-        public static BaseProjectManager? CreateProjectManager(PackageURL purl, IHttpClientFactory? httpClientFactory = null, string? destinationDirectory = null)
+        public delegate BaseProjectManager? ConstructProjectManager(string destinationDirectory = ".");
+
+        /// <summary>
+        /// The dictionary of project managers.
+        /// </summary>
+        private readonly Dictionary<string, ConstructProjectManager> _projectManagers;
+
+        /// <summary>
+        /// Initializes a new instance of a <see cref="ProjectManagerFactory"/>.
+        /// </summary>
+        /// <param name="overrideManagers">Use a custom set of managers to override the defaults with.</param>
+        /// <param name="httpClientFactory">The <see cref="IHttpClientFactory"/> to use in the project managers.</param>
+        public ProjectManagerFactory(Dictionary<string, ConstructProjectManager> overrideManagers, IHttpClientFactory? httpClientFactory = null) : this(httpClientFactory)
         {
-            if (projectManagers.Count == 0)
-            {
-                projectManagers.AddRange(typeof(BaseProjectManager).Assembly.GetTypes()
-               .Where(type => type.IsSubclassOf(typeof(BaseProjectManager))));
-            }
-            // Use reflection to find the correct package management class
-            Type? downloaderClass = projectManagers
-               .Where(type => type.Name.Equals($"{purl.Type}ProjectManager",
-                                               StringComparison.InvariantCultureIgnoreCase))
-               .FirstOrDefault();
-
-            if (downloaderClass != null)
+            overrideManagers.ToList().ForEach(pair => SetManager(pair.Key, pair.Value));
+        }
+
+        /// <summary>
+        /// Initializes a new instance of a <see cref="ProjectManagerFactory"/>.
+        /// </summary>
+        /// <param name="httpClientFactory">The <see cref="IHttpClientFactory"/> to use in the project managers.</param>
+        public ProjectManagerFactory(IHttpClientFactory? httpClientFactory = null)
+        {
+            _projectManagers = CreateDefaultManagers(httpClientFactory);
+        }
+
+        public bool UnsetManager(string lookup)
+        {
+            return _projectManagers.Remove(lookup);
+        }
+
+        private void SetManager(string lookup, ConstructProjectManager generator)
+        {
+            _projectManagers[lookup] = generator;
+        }
+
+        public void ClearManagers()
+        {
+            _projectManagers.Clear();
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="httpClientFactoryParam">The <see cref="IHttpClientFactory"/> to use in the project managers.</param>
+        /// <returns>A dictionary of the default managers with associated constructors.</returns>
+        private static Dictionary<string, ConstructProjectManager> CreateDefaultManagers(IHttpClientFactory? httpClientFactoryParam = null)
```

Comment:
This needs to be public if we remove the specific get/set methods so you can modify the default set and then pass it in.

Reply:
I removed the UnsetManager and ClearManagers methods. I can't think of a case where we would need either of those, and I also can't think of a case where we would need a getter outside of the `ProjectManagerFactory` class.